pkg/frontend/querymiddleware/codec.go

```diff
@@ -112,7 +116,46 @@ type Response interface {
 	GetHeaders() []*PrometheusResponseHeader
 }
 
-type prometheusCodec struct{}
+type prometheusCodecMetrics struct {
+	duration *prometheus.HistogramVec
+	size     *prometheus.HistogramVec
+}
+
+func newPrometheusCodecMetrics(registerer prometheus.Registerer) *prometheusCodecMetrics {
+	factory := promauto.With(registerer)
+	ms := 1.0 / 1000
+	us := ms / 1000
+	ns := us / 1000
+	kb := 1024.0
+	mb := 1024 * kb
+
+	return &prometheusCodecMetrics{
+		duration: factory.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: "cortex",
+			Name:      "frontend_query_payload_codec_duration_seconds",
+			Help:      "Total time spent encoding or decoding query result payloads, in seconds.",
+			// We need a wide range of buckets here because encoding protobuf payloads can take as little as 80ns, and decoding very large JSON payloads can take many milliseconds.
+			Buckets: []float64{50 * ns, 100 * ns, 500 * ns, 1 * us, 5 * us, 10 * us, 25 * us, 50 * us, 100 * us, 250 * us, 500 * us, 1 * ms, 2.5 * ms, 5 * ms, 10 * ms, 20 * ms},
+		}, []string{"operation", "format"}),
+		size: factory.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: "cortex",
+			Name:      "frontend_query_payload_codec_payload_bytes",
+			Help:      "Total size of encoded or decoded query result payloads, in bytes.",
+			// Protobuf payloads can be as small as a few bytes, and payloads in either format with a large number of labels can easily be 30+ MB.
+			Buckets: []float64{8, 16, 32, 256, 512, kb, 32 * kb, 256 * kb, 512 * kb, mb, 8 * mb, 16 * mb, 32 * mb, 64 * mb, 256 * mb},
```

Comment:
Is there value in differentiating between 8, 16, and 32 bytes? It seems like a minimum of 256 would be fine to bucket all "small" payloads.

Reply:
I'm expecting a lot of payloads that are 11 bytes (smallest possible protobuf payload) and 63 bytes (smallest possible JSON payload). These payloads contain no results and account for ~49% of queries in my sample set from our ops cluster, so I wanted to have them end up in different buckets to payloads with results.

Reply:
I see what you mean Charles, but does it really matter for us? Why do we even care whether it's 8 or 16 bytes? Why don't we just care about the order of magnitude here, so maybe starting from 256KB?

Reply:
I've tweaked the buckets in 6c41f95. We now have a set of 10 exponentially increasing buckets from 1 KB to 512 MB.